go-controller/pkg/ovn/master_test.go

```diff
@@ -1227,3 +1227,261 @@ var _ = ginkgo.Describe("Gateway Init Operations", func() {
+					Annotations: map[string]string{
+						"k8s.ovn.org/node-subnets": `{"default": ["172.16.5.0/24","2000:2::/24"]}`,
+					},
+				},
+			},
+			wantStr:   []string{"172.16.5.0/24"},
+			allocated: 0,
+			wantErr:   false,
+		},
+		{
+			name:          "dual stack cluster to single IPv6",
+			networkRanges: []string{"2001:db2::/56"},
+			networkLen:    64,
+			configIPv4:    false,
+			configIPv6:    true,
+			node: &v1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "testnode",
+					Annotations: map[string]string{
+						"k8s.ovn.org/node-subnets": `{"default": ["172.16.5.0/24","2001:db2:1:2:3:4::/64"]}`,
+					},
+				},
+			},
+			wantStr:   []string{"2001:db2:1:2:3:4::/64"},
+			allocated: 0,
+			wantErr:   false,
+		},
+		{
+			name:          "new node, OVN wrong configuration: IPv4 only cluster but IPv6 range",
+			networkRanges: []string{"2001:db2::/64"},
+			networkLen:    112,
+			configIPv4:    true,
+			configIPv6:    false,
+			node: &v1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:        "testnode",
+					Annotations: map[string]string{},
+				},
+			},
+			wantStr:   []string{},
+			allocated: 0,
+			wantErr:   true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// create cluster config
+			config.IPv4Mode = tt.configIPv4
+			config.IPv6Mode = tt.configIPv6
+			// create fake OVN controller
+			stopChan := make(chan struct{})
+			defer close(stopChan)
+			kubeFakeClient := fake.NewSimpleClientset()
+			egressFirewallFakeClient := &egressfirewallfake.Clientset{}
+			crdFakeClient := &apiextensionsfake.Clientset{}
+			egressIPFakeClient := &egressipfake.Clientset{}
+			fakeClient := &util.OVNClientset{
+				KubeClient:           kubeFakeClient,
+				EgressIPClient:       egressIPFakeClient,
+				EgressFirewallClient: egressFirewallFakeClient,
+				APIExtensionsClient:  crdFakeClient,
+			}
+			f, err := factory.NewMasterWatchFactory(fakeClient)
+			clusterController := NewOvnController(fakeClient, f, stopChan,
+				addressset.NewFakeAddressSetFactory(), ovntest.NewMockOVNClient(goovn.DBNB),
+				ovntest.NewMockOVNClient(goovn.DBSB), record.NewFakeRecorder(0))
+
+			// configure the cluster allocators
+			for _, subnetString := range tt.networkRanges {
+				_, subnet, err := net.ParseCIDR(subnetString)
+				if err != nil {
+					t.Fatalf("Error parsing subnet %s", subnetString)
+				}
+				clusterController.masterSubnetAllocator.AddNetworkRange(subnet, tt.networkLen)
+			}
+			// test network allocation works correctly
+			got, allocated, err := clusterController.allocateNodeSubnets(tt.node)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Controller.addNode() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			want := make([]*net.IPNet, len(tt.wantStr))
+			for i, w := range tt.wantStr {
+				_, want[i], err = net.ParseCIDR(w)
+				if err != nil {
+					t.Fatalf("Error parsing subnet %s", w)
+				}
+			}
+			if !apiequality.Semantic.DeepEqual(got, want) {
```

Comment:
reflect.DeepEqual doesn't consider nil = nil

Reply:
no, it does. Your issue here is probably that `got` is `nil` but `want` is a 0-length array, which is not exactly the same thing (though for kubernetes API semantic purposes, it is).

Reply:
these things always confuse me, let me fix it then

Reply:
fixed, I was using a make some lines above to created the `want` slice, so always was never nil and it failed